r/R/csv.R

```diff
@@ -510,32 +510,55 @@ CsvReadOptions$create <- function(use_threads = option_use_threads(),
   options
 }
 
-readr_to_csv_write_options <- function(include_header = TRUE,
+readr_to_csv_write_options <- function(col_names = TRUE,
                                        batch_size = 1024L,
-                                       na = "") {
+                                       delim = ",",
+                                       na = "",
+                                       eol = "\n",
+                                       quote = "Needed") {
```

Comment:
The valid values for `quote` in `readr::write_delim()` are `("needed", "all", "none")`, so do we need to allow those values here and then convert them to the Arrow-valid version in the body of this function?

Reply:
I think it's cleaner to only support the Arrow version, otherwise you need to validate and support both options:

```r
write_dataset(ds, path, quote = "Needed")

write_delim_dateset(ds, path, quote = "needed")
```

Also, I'm not sure the behaviour of the `needed` options are identical. Specifically Arrow only encloses for quotes, whereas readr encloses for a delimiter, quote, or newline. The arrow options are [here](https://github.com/apache/arrow/blob/9b3bf08f78f690605394ad3815177ece931f06d3/cpp/src/arrow/csv/options.h#L173-L186): 

```c++
/// \brief Quoting style for CSV writing
enum class ARROW_EXPORT QuotingStyle {
  /// Only enclose values in quotes which need them, because their CSV rendering can
  /// contain quotes itself (e.g. strings or binary values)
  Needed,
  /// Enclose all valid values in quotes. Nulls are not quoted. May cause readers to
  /// interpret all values as strings if schema is inferred.
  AllValid,
  /// Do not enclose any values in quotes. Prevents values from containing quotes ("),
  /// cell delimiters (,) or line endings (\\r, \\n), (following RFC4180). If values
  /// contain these characters, an error is caused when attempting to write.
  None
};

```

The `readr::write_*` options for `quote` [are](https://readr.tidyverse.org/reference/write_delim.html): 

* needed - Values are only quoted if needed: if they contain a delimiter, quote, or newline.
* all - Quote all fields.
* none - Never quote fields.

Reply:
This is a tricky one; the aim of supporting the `readr` parameters is so that folks can use the `arrow` functions as drop-in replacements for their `readr` etc code, and so it follows that we'd support both the `readr` parameter names and possible values.  For example, we support things like compact column specification in `read_csv_arrow()` even though Arrow itself doesn't support this.

It also feels a bit divergent to have parameter values which are using Pascal Case here as well when everything else is in snake case.

@nealrichardson I'd be keen to get your thoughts here given you were involved in a lot of the design of the earlier functions.

Reply:
I've went ahead and added the readr options and updated the docs and tests to make sure we are testing the output. Notably[ Arrow quotes all strings and binaries](https://github.com/apache/arrow/blob/a7d5217937ca8b8be77995910f672c5e61e9d699/cpp/src/arrow/csv/writer.h#L31-L39), not just in cases where there is a quote or delimiter in the string:

```c++
// Functionality for converting Arrow data to Comma separated value text.
// This library supports all primitive types that can be cast to a StringArrays.
// It applies to following formatting rules:
//  - For non-binary types no quotes surround values.  Nulls are represented as the empty
//  string.
//  - For binary types all non-null data is quoted (and quotes within data are escaped
//  with an additional quote).
//    Null values are empty and unquoted.
```

I've learned more than I expected about the Arrow CSV writer through this PR 😅. @thisisnic @paleolimbot let me know if there's anything else on this one.